### ISSUE TITLE
Recreate investigation - Blocklist blocks not found by recreate?

### ISSUE THREAD

**Issue by tygill** [duplicati/duplicati#3210]
tl;dr: Block for a blocklist not present in database after recreate from remote - original database lists this block as living in an `Index` volume rather than a `Blocks` volume (but the indicated `Index` volume contains no mention of the block's hash).

Several weeks ago, I attempted to recreate my database for a backup after migrating to a new backend (from `OneDrive` to `OneDrive v2`). However, after the repair finished (which ended up processing all `dlist`, `dindex`, and `dblock` files on the remote), the new database was still not usable - when running the `LocalDatabase.VerifyConsistency` check, it would fail with an error stating: `Found inconsistency in the following files while validating database`, and then listed several files which had non-zero `actual` sizes, but a `dbsize` of zero. ([This](https://forum.duplicati.com/t/found-inconsistency-in-the-following-files-while-validating-database/525) forum post and bug #3085 seem to reference this same error case.)

_In that forum post and bug, it is mentioned that deleting the BlocksetIDs associated with the failing files can solve this error, but when I tried that, it just changed the error to one indicating that certain filesets were now missing files, which makes sense as removing anything referencing those Blocksets effectively removes those files from any filesets, so the only way I see to making that workaround actually work would be to also remove all now-broken filesets and start over with only the existing blocks._

Since it looked like a bug that was hitting several other people, I decided to try to dig in to see if I could track down exactly what was causing the bug - was something in the repair / recreate process broken, or was something in my remote backup corrupted / incorrect. In that process, I found a few things that seem quite suspicious, though they have to do with parts of Duplicati that I'm not currently very familiar with, so I'm hoping that by walking through this process I can get some pointers on what some of this might mean. I have a copy of my original SQLite database, as well as the recreated one, which I've used to try to understand what might have happened.

The first thing I noticed was that in the recreated database, there were no entries in either the `BlocklistHash` or `BlocksetEntry` tables for the BlocksetIDs of the inconsistent files, though the `Blockset` table did have the entry (which listed the correct `Length` and `FullHash`). This is ultimately the cause for the consistency error - the length of each file is computed by summing the sizes of each block that makes up the corresponding blockset, but since there are no blockset entries mapping to actual blocks, it computes a `dbsize` of zero, which doesn't match the value in the `Blockset` table.

To figure out why that was, I pulled the raw backup data and started digging through it. (There was lots of mapping BlocksetIDs in the new database to hashes and lining them up to the BlocksetIDs in the original database so I could find which remote index and block volumes to look at...) In this process, I learned something that I hadn't previously seen documented: `dindex` files contain a copy of each block from the corresponding `dblock` file that represents a blocklist (in the `list` folder). I assume this duplication is so that the database can be rebuilt without needing the `dblock` files, as the blocksets can be recreated entirely from the `dindex` file contents this way. (Otherwise, I assume the `dblock` could be downloaded and every block that happens to be a size that corresponds to a valid blocklist could be assumed to be a blocklist, which the `dlist` files could then reference back to - I'm not sure whether Duplicati currently does this kind of repair today though.)

Since every blocklist is a block, I used the original database to find the block with the expected hash and then which volume contained that block. Strangely, the database said that an `Index` volume was the one that contained it. However, I looked at that `dindex` file myself, and there was no sign of that hash value (either as a blocklist block in the `list` folder, or as a block entry in the `dblock` description of the `vol` folder). Interestingly, the new, recreated database didn't have an entry for a block with this hash in the `Block` table either.

_As a sanity check, I did the same exercise for a blocklist for a file which was fine, and that block was listed as being in a `Blocks` volume, as expected, and the corresponding `dindex` file both had the full block in the `list` folder and listed the hash and size in the `dblock` index._

Base on this, it seems like the block containing this blocklist was listed in the database without actually being put into a remote volume (as the recreate never found it). How is that possible? That seems like a pretty major issue - even if this was ultimately caused by a crash or other exceptional situation, Duplicati should be able to handle it and recover from it by determining that the block hasn't actually been uploaded.

I checked the original database by running the following script, and it shows that ~10000 blocks are listed as living in `Index` volumes. The same query on my recreated database returned 0 blocks.

`SELECT COUNT() AS Count, Remotevolume.ID AS VolumeID FROM Remotevolume INNER JOIN Block ON Block.VolumeID == Remotevolume.ID WHERE Remotevolume.Type == "Index" GROUP BY VolumeID`

**Comment by warwickmm** [comment]
I believe this was fixed by pull request #3225.
